### ISSUE TITLE
Active Test-Time Adaptation: Theoretical Analyses and An Algorithm

### ISSUE THREAD

**Issue by zhuoxiao-chen** [zhuoxiao-chen/please_read_paper_every_day#61]
<img width="1330" alt="image" src="https://github.com/user-attachments/assets/685780b0-c108-4083-a281-d0ca14d0cb73">

### Paper Link
https://openreview.net/forum?id=YHUGlwTzFB

### Code Link
https://github.com/divelab/ATTA

## Why AL for TTA?

Traditional FTTA aims at adapting a pre-trained model to streaming test-time data from diverse domains under unsupervised settings. However, recent works (Lin et al., 2022; Pearl, 2009) prove that **it is theoretically infeasible to achieve OOD generalization without extra information** such as environment partitions. Since utilizing environment partitions requires heavy pretraining (DG), contradicting the nature of TTA, we are motivated to incorporate extra information in a different way, i.e., integrating **a limited number of labeled test-time samples** to alleviate distribution shifts, following the active learning (AL) paradigm (Settles, 2009).

## SimATTA framework

<img width="1300" alt="image" src="https://github.com/user-attachments/assets/3081b4b9-4caa-4ca2-aec7-c71abb015b62">

-  $(x, y)$ be a labeled sample and $f(\cdot; \theta)$ be our neural network, where $\hat{y} = f(x; \theta)$ and $\theta$ represents the parameters. 
- We have a model **pre-trained on source domains** with the pre-trained parameters $\phi$. We initialize model parameters as $\theta(0) = \phi$ and aim to adapt the model $f(\cdot; \theta)$ in real-time. 
- During the test phase, the model continuously predicts labels for streaming-in test data and concurrently gets fine-tuned. We perform sample selection to enable active learning. 
- At each time step $t$, we first partition unlabeled test samples $U_{\text{te}}(t)$ into **high entropy** and **low entropy** datasets, $U_h(t)$ and $U_l(t)$, using an **entropy threshold**. 
- The source-pretrained model $f(\cdot; \phi)$ is frozen to predict pseudo labels for low entropy data. We obtain labeled low-entropy data $D_l(t)$ by labeling $U_l(t)$ with $f(\cdot; \phi)$ and combining it with $D_l(t-1)$. 
- To select of high-entropy samples for active labeling, we design an **incremental clustering** sample selection technique to reduce sample redundancy and increase distribution coverage. The incremental clustering algorithm outputs the labeled test samples $D_h(t)$, also referred to as **anchors**, given $D_h(t-1)$ and $U_h(t)$. After sample selection, the model undergoes test-time training using the labeled test anchors $D_h(t)$ and pseudo-labeled source-like anchors $D_l(t)$. 

### Incremental Clustering
- We first extract all sample features using the model from the previous step $f(\cdot; \theta(t-1))$, and then cluster these weighted features using weighted K-Means.
- In each iteration, we apply weighted K-Means for previously selected anchors $D_{\text{anc}}$ and the new streaming-in unlabeled data $U_{\text{new}}$. 
- We first extract all sample features using the model from the previous step $f(\cdot; \theta(t-1))$, and then cluster these weighted features. 
- After clustering, clusters including old anchors are old clusters, while clusters only containing new samples are newly formed ones. For each new cluster, we select the centroid-closest sample as the new anchor to store. 
- With incremental clustering, **although we can control the number of clusters in each iteration, we cannot control the number of new clusters/new anchors**. This indirect control makes the increase of **new anchors adaptive to the change of distributions, but it also leads to indirect budget control**. Therefore, in experimental studies, we set the budget limit $B$, but the actual anchor budget will not reach this limit. 
## Implementation 

Our implementation largely involves straightforward hyperparameter settings. The higher entropy bound $e_h = 10^{-2}$ should exceed the lower entropy bound $e_l$, but equal values are acceptable. Empirically, the lower entropy bound $e_l$ can be set to $10^{-3}$ for VLCS and Office-Home, or $10^{-4}$ for PACS. The choice of $e_l$ is largely dependent on the number of source-like samples obtained. A lower $e_l$ may yield higher-accuracy low-entropy samples, but this could lead to unstable training due to sample scarcity.

Regarding the management of budgets, numerous strategies can be adopted. In our experiments, we utilized a simple hyperparameter $k$, varying from 1 to 3, to regulate the increasing rate of budget consumption. This strategy is fairly elementary and can be substituted by any adaptive techniques.

## Datasets and Baselines
- For each dataset, we designate one domain as the source domain and arrange the samples from the other domains to form the test data stream. We choose P in PACS and C in VLCS as source domains.
- For **DomainBed** datasets, we adopt two stream order strategies. The first order uses a domain-wise data stream, i.e., we finish streaming samples from one domain before starting streaming another domain. The second order is random, where we shuffle samples from all target domains and partition them into four splits 1, 2, 3, and 4.
- We report the **real-time adaptation accuracy** results for each split of the data stream, as well as the **accuracy on each domain after all adaptations** through the data stream (under “post-adaptation” columns).
- **Enhanced TTA** is built on TTA with access to extra random sample labels. TTA baselines are further fine-tuned with these random samples.
- We also use ADA methods for comparisons. ADA baselines have access to all samples in the pre-collected target datasets but not source domain data, whereas our method can only access the streaming test data.
- For baseline models, we start with the common source-only models, which either utilize pre-calculated batch statistics (BN w/o adapt) or test batch statistics (BN w/ adapt).

### TTA Comparisons on PACS and VLCS
<img width="1446" alt="image" src="https://github.com/user-attachments/assets/efac5726-16a6-42b1-8ba5-0758b65250bd">

- "step=10" indicates 10 test-time training updates, i.e., 10 gradient backpropagation iterations.

### Enhanced TTA Comparisons on TinyImageNet-C
<img width="1273" alt="image" src="https://github.com/user-attachments/assets/a90d7c44-0a0d-4775-b843-141d1df5d34c">

- The classifier of ResNet18-BN is pre-adapted to the brightness corruption (source domain) before test-time adapting.
- SimATTA’s label budget is around 4,000, while all other TTA methods have budget 4,500 for randomly selected labeled samples.
- **Simply accessing labeled samples cannot benefit TTA methods to match ATTA**. 

### Comparisons to ADA Baselines
<img width="620" alt="image" src="https://github.com/user-attachments/assets/6d841a81-547f-48ca-ad19-52600f24be9b">

### Ablation Study

<img width="851" alt="image" src="https://github.com/user-attachments/assets/2b795300-82cb-4a35-a27d-71f211090990">

- SimATTA (LE=1, IC=1) comprehensively outperforms other variants on both datasets, different streams, and different budgets.
- Variants without low-entropy training (LE=0, IC=0/1) easily fail to produce stable results
- IC has its **unique advantage on domain-wise streams** where distributions change abruptly instead of random streams. Therefore, compared to PACS’s domainwise stream results, the reason for the smaller performance improvement of SimATTA over the variant (LE=1, IC=0) on VLCS’s domain-wise stream is that images in VLCS are all photos that do not include those severe style shifts in PACS (i.e., art, cartoons, and sketches). That is, when the shift is not severe, we don’t need IC to cover very different distributions, and selecting samples using entropy can produce good results. In brief, IC is extraordinary for severe distribution shifts and quick adaptation.
